module/zfs/dmu.c

```diff
@@ -1290,515 +1879,502 @@ dmu_write_uio_dnode(dnode_t *dn, uio_t *uio, uint64_t size, dmu_tx_t *tx)
 
-	return (arc_loan_buf(db->db_objset->os_spa, B_FALSE, size));
+	if (size == 0)
+		return;
+
+	VERIFY0(dmu_buf_hold_array(os, object, offset, size,
+	    FALSE, FTAG, &numbufs, &dbp));
+
+	for (i = 0; i < numbufs; i++) {
+		dmu_buf_t *db = dbp[i];
+
+		dmu_buf_will_not_fill(db, tx);
+	}
+	dmu_buf_rele_array(dbp, numbufs, FTAG);
 }
 
-/*
- * Free a loaned arc buffer.
- */
 void
-dmu_return_arcbuf(arc_buf_t *buf)
+dmu_write_embedded(objset_t *os, uint64_t object, uint64_t offset,
+    void *data, uint8_t etype, uint8_t comp, int uncompressed_size,
+    int compressed_size, int byteorder, dmu_tx_t *tx)
 {
-	arc_return_buf(buf, FTAG);
-	arc_buf_destroy(buf, FTAG);
+	dmu_buf_t *db;
+
+	ASSERT3U(etype, <, NUM_BP_EMBEDDED_TYPES);
+	ASSERT3U(comp, <, ZIO_COMPRESS_FUNCTIONS);
+	VERIFY0(dmu_buf_hold_noread(os, object, offset,
+	    FTAG, &db));
+
+	dmu_buf_write_embedded(db,
+	    data, (bp_embedded_type_t)etype, (enum zio_compress)comp,
+	    uncompressed_size, compressed_size, byteorder, tx);
+
+	dmu_buf_rele(db, FTAG);
 }
 
-/*
- * When possible directly assign passed loaned arc buffer to a dbuf.
- * If this is not possible copy the contents of passed arc buf via
- * dmu_write().
- */
-int
-dmu_assign_arcbuf_by_dnode(dnode_t *dn, uint64_t offset, arc_buf_t *buf,
+void
+dmu_redact(objset_t *os, uint64_t object, uint64_t offset, uint64_t size,
     dmu_tx_t *tx)
 {
-	dmu_buf_impl_t *db;
-	objset_t *os = dn->dn_objset;
-	uint64_t object = dn->dn_object;
-	uint32_t blksz = (uint32_t)arc_buf_lsize(buf);
-	uint64_t blkid;
+	int numbufs, i;
+	dmu_buf_t **dbp;
 
-	rw_enter(&dn->dn_struct_rwlock, RW_READER);
-	blkid = dbuf_whichblock(dn, 0, offset);
-	db = dbuf_hold(dn, blkid, FTAG);
-	if (db == NULL)
-		return (SET_ERROR(EIO));
-	rw_exit(&dn->dn_struct_rwlock);
+	VERIFY0(dmu_buf_hold_array(os, object, offset, size, FALSE, FTAG,
+	    &numbufs, &dbp));
+	for (i = 0; i < numbufs; i++)
+		dmu_buf_redact(dbp[i], tx);
+	dmu_buf_rele_array(dbp, numbufs, FTAG);
+}
+
+#ifdef _KERNEL
+static int
+dmu_rw_uio_direct(dnode_t *dn, uio_t *uio, uint64_t size,
+    dmu_tx_t *tx, boolean_t read)
+{
+	uint_t numpages;
+	abd_t *data;
+	int err;
 
 	/*
-	 * We can only assign if the offset is aligned, the arc buf is the
-	 * same size as the dbuf, and the dbuf is not metadata.
+	 * All Direct IO requests must be PAGE_SIZE aligned
 	 */
-	if (offset == db->db.db_offset && blksz == db->db.db_size) {
-		dbuf_assign_arcbuf(db, buf, tx);
-		dbuf_rele(db, FTAG);
-	} else {
-		/* compressed bufs must always be assignable to their dbuf */
-		ASSERT3U(arc_get_compression(buf), ==, ZIO_COMPRESS_OFF);
-		ASSERT(!(buf->b_flags & ARC_BUF_FLAG_COMPRESSED));
+	ASSERT(IO_PAGE_ALIGNED(uio->uio_loffset, size));
 
-		dbuf_rele(db, FTAG);
-		dmu_write(os, object, offset, blksz, buf->b_data, tx);
-		dmu_return_arcbuf(buf);
+	numpages = size / PAGE_SIZE;
+	zfs_page_p *pages =
+	    kmem_alloc(numpages * sizeof (zfs_page_p), KM_SLEEP);
+
+	err = zfs_uio_get_user_pages(uio, pages, numpages,
```

Comment:
We've discussed this directly, but I'll add this here for other reviewers. If possible, I'd like to see the page get() operations happen closer to the top of the stack in `{read,write}_iter` (assuming that covers all bases) and use the `struct bio_vec` pointer present in the Linux uio (for FreeBSD I'll have to add some extra aliasing in the synchronous case) to enable code sharing between this and the async I/O (io_uring, aio, io_submit, etc) interface to async dmu.

Reply:
I agree, this would preferable and seems pretty straight forward on the surface.  I'd suggest adding a pointer for the pages returned by `get_user_pages()` to the Linux definition of the `uio_t`.  Then get and put these pages in the `zpl_{read,write}_iter` functions.  Moving the page pointer in to the `uio_t` should also allow us to get any page handling out of the common code and remove the need for the `zfs_page_p`.  It looks like you'll just need to change `abd_alloc_from_pages()` to expect a `uio_t` instead of pages (and rename it of course).

You should be able to use `iov_iter_get_pages_alloc()` for this all the way back to the 3.16 kernel.  Or in other words it should fine to use in the entire `HAVE_VFS_RW_ITERATE` block of `zpl_file.c`.  For the !HAVE_VFS_RW_ITERATE case `get_user_pages()` can be used with the iovecs to get the pages.  One additional upside is that we won't need to provide our own functions for this and can shed some of the compatibility code.  We should still keep the checks to verify those two functions exist and take the expected arguments.

Reply:
We addressed this in the current push.